lib/response-types/code-response-type.js

```diff
@@ -4,33 +4,193 @@
  * Module dependencies.
  */
 
+var _ = require('lodash');
+var AbstractResponseType = require('./abstract-response-type');
 var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var InvalidScopeError = require('../errors/invalid-scope-error');
+var UnauthorizedClientError = require('../errors/unauthorized-client-error');
+var OAuthError = require('../errors/oauth-error');
+var Promise = require('bluebird');
+var ServerError = require('../errors/server-error');
+var is = require('../validator/is');
+var tokenUtil = require('../utils/token-util');
+var util = require('util');
 var url = require('url');
 
 /**
  * Constructor.
  */
 
-function CodeResponseType(code) {
-  if (!code) {
-    throw new InvalidArgumentError('Missing parameter: `code`');
+function CodeResponseType(options) {
+  options = options || {};
+
+  if (!options.authorizationCodeLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `authorizationCodeLifetime`');
+  }
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
   }
 
-  this.code = code;
+
+  if (!options.model.saveAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveAuthorizationCode()`');
+  }
+
+  this.allowEmptyState = options.allowEmptyState;
+  this.authorizationCodeLifetime = options.authorizationCodeLifetime;
+
+  AbstractResponseType.call(this, options);
 }
 
 /**
- * Build redirect uri.
+ * Inherit prototype.
+ */
+
+util.inherits(CodeResponseType, AbstractResponseType);
+
+/**
+ * Handle `code` response type.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+ */
+
+CodeResponseType.prototype.handle = function (request, client, user) {
+  if (!request) {
+    throw new InvalidArgumentError('Missing parameter: `request`');
+  }
+
+  if (!client) {
+    throw new InvalidArgumentError('Missing parameter: `client`');
+  }
+
+  if (!_.includes(client.grants, 'authorization_code')) {
+    throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+  }
+
+  var fns = [
+    this.generateAuthorizationCode(),
+  ];
+  var uri = this.getRedirectUri(request, client);
+  var state;
+
+  return Promise.all(fns)
+    .bind(this)
+    .spread(function (authorizationCode) {
+      var scope = this.getScope(request);
+      var expiresAt = this.getAuthorizationCodeLifetime();
+
+      return this.saveAuthorizationCode(authorizationCode, expiresAt, scope, client, uri, user);
+    })
+    .tap(function () {
+      state = this.getState(request);
```

Comment:
Any reason why this can't be done when declaring the variable?

Reply:
Yes. We want to catch `getState()` errors to include them on the `redirectUri` but we need to declare the variable outside so we don't need to change the returns messages.